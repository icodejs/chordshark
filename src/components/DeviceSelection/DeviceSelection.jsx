import React, { Component } from 'react';
import PropTypes from 'prop-types';
import webmidi from 'webmidi';

import './DeviceSelection.scss';

class DeviceSelection extends Component {
  componentDidMount() {
    this.props.onReceivedMidiInputs([...webmidi.inputs]);
  }

  handleDeviceSelection = event => {
    const { value: selectedDeviceName } = event.target;
    const input = webmidi.getInputByName(selectedDeviceName);
    const output = webmidi.getOutputByName(selectedDeviceName);

    this.props.onDeviceSelection({ input, output, selectedDeviceName });

    // changing the midi device should remove listener from the prev device
  };

  render() {
    const { midiInputs } = this.props;

    if (midiInputs && midiInputs.length === 0) {
      return <div className="device-selection">No Midi devices found.</div>;
    }

    return (
      <form className="device-selection">
        <label>
          <select
            value={this.props.selectedDeviceName}
            onChange={this.handleDeviceSelection}
          >
            <option value="-1">Choose device</option>
            {this.props.midiInputs.map(({ id, manufacturer, name }) => {
              return (
                <option key={id} value={name}>
                  {`${manufacturer} ${name}`}
                </option>
              );
            })}
          </select>
        </label>
      </form>
    );
  }
}

DeviceSelection.propTypes = {
  onReceivedMidiInputs: PropTypes.func.isRequired,
  onDeviceSelection: PropTypes.func.isRequired,
  midiInputs: PropTypes.array,
};

DeviceSelection.defaultProps = {
  midiInputs: [],
};

export default DeviceSelection;
