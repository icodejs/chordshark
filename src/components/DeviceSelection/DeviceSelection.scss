.device-selection {
  display: inline-block;
}
