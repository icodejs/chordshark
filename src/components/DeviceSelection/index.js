import DeviceSelection from './DeviceSelection';
export default DeviceSelection;
