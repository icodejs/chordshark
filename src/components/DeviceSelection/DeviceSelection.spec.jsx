import React from 'react';
import renderer from 'react-test-renderer';

import DeviceSelection from './DeviceSelection';

describe('DeviceSelection', () => {
  describe('render', () => {
    it('should render component', () => {
      const props = {
        onReceivedMidiInputs: () => {},
        onDeviceSelection: () => {},
      };
      const json = renderer.create(<DeviceSelection {...props} />).toJSON();
      expect(json).toMatchSnapshot();
    });
  });
});
