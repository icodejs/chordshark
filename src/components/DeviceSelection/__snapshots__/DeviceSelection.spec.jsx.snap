// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeviceSelection render should render component 1`] = `
<div
  className="device-selection"
>
  No Midi devices found.
</div>
`;
